controller-server/src/main/java/com/yahoo/vespa/hosted/controller/application/pkg/ApplicationPackageValidator.java

```diff
@@ -59,6 +59,16 @@ public void validate(Application application, ApplicationPackage applicationPack
         validateEndpointChange(application, applicationPackage, instant);
         validateCompactedEndpoint(applicationPackage);
         validateSecurityClientsPem(applicationPackage);
+        validateDeprecatedElements(applicationPackage);
+    }
+
+    /** Verify that deployment spec does not use elements deprecated on a major version older than compile version */
+    private void validateDeprecatedElements(ApplicationPackage applicationPackage) {
+        for (var deprecatedElement : applicationPackage.deploymentSpec().deprecatedElements()) {
+            if (applicationPackage.compileVersion().isEmpty()) continue;
```

Comment:
Hmm ... so should we try to set a compile version for non-java projects as well, then? Assign the controller version at the time of submission, or use the one from deployment.xml if set?

Reply:
Yes, that makes sense. I'll fix.